app/src/main/java/com/google/jetpackcamera/MainActivity.kt

```diff
@@ -101,13 +107,67 @@ class MainActivity : ComponentActivity() {
                             modifier = Modifier.fillMaxSize(),
                             color = MaterialTheme.colorScheme.background
                         ) {
-                            JcaApp(onPreviewViewModel = { previewViewModel = it })
+                            JcaApp(
+                                onPreviewViewModel = { previewViewModel = it },
+                                contentResolver = contentResolver,
+                                imageCaptureUri = uri,
+                                onImageCapture = { event ->
+                                    val shouldFinishWithResult = isExternalImageCapture()
+                                    when (event) {
+                                        is CameraUseCase.ImageCaptureEvent.ImageSaved -> {
+                                            if (shouldFinishWithResult) {
+                                                setResult(RESULT_OK)
+                                                finish()
+                                            }
+                                        }
+
+                                        is CameraUseCase.ImageCaptureEvent.ImageCaptureError -> {
+                                            if (shouldFinishWithResult) {
+                                                setResult(RESULT_CANCELED)
+                                                finish()
+                                            }
+                                        }
+                                    }
+                                }
+                            )
                         }
                     }
                 }
             }
         }
     }
+
+    private fun getImageCaptureUriIfExists(): Uri? {
+        val isExternalImageCapture = isExternalImageCapture()
+        val uri = if (!isExternalImageCapture || intent.extras == null ||
+            !intent.extras!!.containsKey(MediaStore.EXTRA_OUTPUT)
```

Comment:
The [contract](https://developer.android.com/reference/android/provider/MediaStore#ACTION_IMAGE_CAPTURE) also says the URI might come from `ClipData` set on the intent, rather than from `EXTRA_OUTPUT`.